microsetta_admin/server.py

```diff
@@ -774,6 +775,191 @@ def scan():
                                       recorded_type)
 
 
+def get_legends(criteria):
+
+    dict = {}
+    hex = 'ABCDEF0123456789'
+    if criteria not in session:
+        status, fields = APIRequest.get(
+            '/api/admin/barcode_query_fields')
+
+        if status == 200:
+            for obj in fields:
+                if criteria == "Type" and obj["label"] == "Sample Site":
+                    for val in obj["values"]:
+                        code = [random.choice(hex) for i in range(6)]
+                        color_code = ['#' + ''.join(code)]
+                        while color_code[0] in dict:
+                            code = [random.choice(hex) for i in range(6)]
+                            color_code = ['#' + ''.join(code)]
+                        dict[obj["values"][val]] = color_code[0]
+
+                if criteria == "Status" and obj["label"] == "Sample Status":
+                    for val in obj["values"]:
+                        code = [random.choice(hex) for i in range(6)]
+                        color_code = ['#' + ''.join(code)]
+                        while color_code[0] in dict:
+                            code = [random.choice(hex) for i in range(6)]
+                            color_code = ['#' + ''.join(code)]
+                        tmp = obj["values"][val].replace(' ', '-').lower()
+                        dict[tmp] = color_code[0]
+
+                if criteria == "Project" and obj["label"] == "Project":
+                    for val in obj["values"]:
+                        if not ('-' in obj["values"][val]):
+                            code = [random.choice(hex) for i in range(6)]
+                            color_code = ['#' + ''.join(code)]
+                            while color_code[0] in dict:
+                                code = [random.choice(hex) for i in range(6)]
+                                color_code = ['#' + ''.join(code)]
+                            dict[obj["values"][val]] = color_code[0]
+
+            if len(dict) > 0:
+                dict['None'] = "#000000"
+                session[criteria] = dict
+    else:
+        dict = session.get(criteria)
+    print(str(dict))
+    return dict
+
+
+def _compose_table(legends, type):
+    output_dict = {}
+    scanned_data = session.get('scan_data')
```

Comment:
Same question about session scope.

Reply:
scan_dta variable in session stores the data read from csv file. This is preserved within the session variables throughout the session. This is necessary requirement. Alternately we could store as cookie on client end.